sdk/storage/Azure.Storage.Blobs/tests/EncryptedBlockBlobClientTests.cs

```diff
@@ -18,25 +25,281 @@ public EncryptedBlockBlobClientTests(bool async)
+                    keyEncryptionKey: mock,
+                    keyResolver: mock);
+
+            return disposable;
+        }
+
+        // TODO this is a copy/paste. fix that.
+        /// <summary>
+        /// Gets and validates a blob's encryption-related metadata
+        /// </summary>
+        /// <param name="metadata">The blob's metadata</param>
+        /// <returns>The relevant metadata.</returns>
+        private EncryptionData GetAndValidateEncryptionData(IDictionary<string, string> metadata)
+        {
+            _ = metadata ?? throw new InvalidOperationException();
+
+            EncryptionData encryptionData;
+            if (metadata.TryGetValue(EncryptionConstants.ENCRYPTION_DATA_KEY, out string encryptedDataString))
+            {
+                using (var stream = new MemoryStream(Encoding.UTF8.GetBytes(encryptedDataString)))
+                {
+                    var serializer = new DataContractJsonSerializer(typeof(EncryptionData));
+                    encryptionData = (EncryptionData)serializer.ReadObject(stream);
+                }
+            }
+            else
+            {
+                throw new InvalidOperationException("Encryption data does not exist.");
+            }
+
+            _ = encryptionData.ContentEncryptionIV ?? throw new NullReferenceException();
+            _ = encryptionData.WrappedContentKey.EncryptedKey ?? throw new NullReferenceException();
+
+            // Throw if the encryption protocol on the message doesn't match the version that this client library
+            // understands
+            // and is able to decrypt.
+            if (EncryptionConstants.ENCRYPTION_PROTOCOL_V1 != encryptionData.EncryptionAgent.Protocol)
+            {
+                throw new ArgumentException(
+                    "Invalid Encryption Agent. This version of the client library does not understand the " +
+                        $"Encryption Agent set on the queue message: {encryptionData.EncryptionAgent}");
+            }
+
+            return encryptionData;
+        }
+
+        private byte[] LocalManualEncryption(byte[] data, byte[] key, byte[] iv)
+        {
+            using (var aesProvider = new AesCryptoServiceProvider() { Key = key, IV = iv })
+            using (var encryptor = aesProvider.CreateEncryptor())
+            using (var memStream = new MemoryStream())
+            using (var cryptoStream = new CryptoStream(memStream, encryptor, CryptoStreamMode.Write))
+            {
+                cryptoStream.Write(data, 0, data.Length);
+                cryptoStream.FlushFinalBlock();
+                return memStream.ToArray();
+            }
+        }
+
+        #endregion
+
+        [TestCase(16)] // a single cipher block
+        [TestCase(14)] // a single unalligned cipher block
+        [TestCase(Constants.KB)] // multiple blocks
+        [TestCase(Constants.KB - 4)] // multiple unalligned blocks
+        [LiveOnly]
+        public void Upload(long dataSize)
```

Comment:
We only have to write Async tests.  The framework magically converts them to sync tests for us (if you call InstrumentClient, etc.).  Sean can give you a rundown.